providers/gcp/resources/discovery.go

```diff
@@ -362,3 +384,127 @@ func cloneConfig(invConf *inventory.Config) *inventory.Config {
 	invConfClone.Discover = &inventory.Discovery{}
 	return invConfClone
 }
+
+func resolveGcr(ctx context.Context, conf *inventory.Config) ([]*inventory.Asset, error) {
+	resolved := []*inventory.Asset{}
+	repository := conf.Host
+
+	log.Debug().Str("registry", repository).Msg("fetch meta information from gcr registry")
+	gcrImages := NewGCRImages()
+	assetList, err := gcrImages.ListRepository(repository, true)
+	if err != nil {
+		log.Error().Err(err).Msg("could not fetch gcr images")
+		return nil, err
+	}
+
+	for i := range assetList {
+		log.Debug().Str("name", assetList[i].Name).Str("image", assetList[i].Connections[0].Host+assetList[i].Connections[0].Path).Msg("resolved image")
+		resolved = append(resolved, assetList[i])
+	}
+
+	return resolved, nil
+}
+
+func NewGCRImages() *GcrImages {
+	return &GcrImages{}
+}
+
+type GcrImages struct{}
+
+func (a *GcrImages) Name() string {
+	return "GCP Container Registry Discover"
+}
+
+// lists a repository like "gcr.io/mondoo-base-infra"
+func (a *GcrImages) ListRepository(repository string, recursive bool) ([]*inventory.Asset, error) {
+	repo, err := name.NewRepository(repository)
+	if err != nil {
+		log.Fatal().Err(err).Str("repository", repository).Msg("could not create repository")
+	}
+
+	auth, err := google.Keychain.Resolve(repo.Registry)
+	if err != nil {
+		log.Fatal().Err(err).Str("repository", repository).Msg("failed to get auth for repository")
+	}
+
+	imgs := []*inventory.Asset{}
+
+	toAssetFunc := func(repo name.Repository, tags *google.Tags, err error) error {
+		if err != nil {
+			return err
+		}
+
+		for digest := range tags.Manifests {
+			repoURL := repo.String()
+			imageUrl := repoURL + "@" + digest
+
+			asset := &inventory.Asset{
+				Connections: []*inventory.Config{
+					{
+						Type: "container-registry",
+						Host: imageUrl,
+					},
+				},
+			}
+			imgs = append(imgs, asset)
+		}
+		return nil
+	}
+
+	// walk nested repos
+	if recursive {
+		err := google.Walk(repo, toAssetFunc, google.WithAuth(auth))
+		if err != nil {
+			return nil, err
+		}
+		return imgs, nil
+	}
+
+	// NOTE: since we're not recursing, we ignore tags.Children
+	tags, err := google.List(repo, google.WithAuth(auth))
+	if err != nil {
+		return nil, err
+	}
+
+	err = toAssetFunc(repo, tags, nil)
+	if err != nil {
+		return nil, err
+	}
+	return imgs, nil
+}
+
+// List uses your GCP credentials to iterate over all your projects to identify potential repos
+func (a *GcrImages) List() ([]*inventory.Asset, error) {
+	assets := []*inventory.Asset{}
+
```

Comment:
why not use the resource directly here instead of calling the api?

Reply:
My understanding is, we do not have these resources.

At the end, `gcp gcr` and `container registry` commands, both create `container.image` resources, but how we get there is different.

`gcp gcr` does not use a container registry API, but the GCP API. We do not have the intermediate resources `gcp.gcr` and `gcp.gcr.repository`. As I understand this provider, we do not care about these.

So, we start by iterating over all repositories which are present in a GCP project. We take these repositories and iterate over all the images which in these repositories. These images end up as resources, but nothing in between.